cranelift/codegen/src/machinst/lower.rs

```diff
@@ -0,0 +1,723 @@
+                        let edge_block = next_bindex;
+                        next_bindex += 1;
+                        edge_blocks_by_inst[inst].push(edge_block);
+                        edge_blocks.push((inst, edge_block, next_bb));
+                    };
+                    for succ in branch_targets(self.f, *bb, inst).into_iter() {
+                        add_succ(succ);
+                    }
+                }
+            }
+        }
+
+        for bb in bbs.iter() {
+            debug!("lowering bb: {}", bb);
+
+            // If this is a return block, produce the return value setup.
+            let last_insn = self.f.layout.block_insts(*bb).last().unwrap();
+            let last_insn_opcode = self.f.dfg[last_insn].opcode();
+            if last_insn_opcode.is_return() {
+                let gen_ret = if last_insn_opcode == Opcode::Return {
+                    GenerateReturn::Yes
+                } else {
+                    debug_assert!(last_insn_opcode == Opcode::FallthroughReturn);
+                    GenerateReturn::No
+                };
+                self.gen_retval_setup(gen_ret);
+                self.vcode.end_ir_inst();
+            }
+
+            // Find the branches at the end first, and process those, if any.
+            let mut branches: SmallVec<[Inst; 2]> = SmallVec::new();
+            let mut targets: SmallVec<[BlockIndex; 2]> = SmallVec::new();
+
+            for inst in self.f.layout.block_insts(*bb).rev() {
+                debug!("lower: inst {}", inst);
+                if edge_blocks_by_inst[inst].len() > 0 {
+                    branches.push(inst);
+                    for target in edge_blocks_by_inst[inst].iter().rev().cloned() {
+                        targets.push(target);
+                    }
+                } else {
+                    // We've reached the end of the branches -- process all as a group, first.
+                    if branches.len() > 0 {
+                        let fallthrough = self.f.layout.next_block(*bb);
+                        let fallthrough = fallthrough.map(|bb| self.vcode.bb_to_bindex(bb));
+                        branches.reverse();
+                        targets.reverse();
+                        debug!(
+                            "lower_branch_group: targets = {:?} branches = {:?}",
+                            targets, branches
+                        );
+                        backend.lower_branch_group(
+                            &mut self,
+                            &branches[..],
+                            &targets[..],
+                            fallthrough,
+                        );
+                        self.vcode.end_ir_inst();
+                        branches.clear();
+                        targets.clear();
+                    }
+
+                    // Only codegen an instruction if it either has a side
+                    // effect, or has at least one use of one of its results.
```

Comment:
Nope, not required; the reasoning behind including it here is that it is intertwined with the operand-matching/merging behavior. For example, an `iadd` can incorporate an `iconst` on an input; it then calls `ctx.merged(input)` which indicates that the use was incorporated directly, updating the refcounts (use counts). This could just as well be handled by a later inst-level DCE, but it seemed cheaper to avoid the lowering logic in the first place (Julian's measurement earlier today was ~4K host x86 instructions spent to lower one target instruction; that should improve but still). If you think we should factor these apart, I'm happy to do that, though!